components/combined_utility/browser/combined_utility_service.cc

```diff
@@ -0,0 +1,103 @@
+/* Copyright (c) 2022 The Brave Authors. All rights reserved.
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#include "brave/components/combined_utility/browser/combined_utility_service.h"
+
+#include "components/keyed_service/content/browser_context_dependency_manager.h"
+#include "content/public/browser/service_process_host.h"
+
+template <>
+inline sandbox::mojom::Sandbox
+content::GetServiceSandboxType<combined_utility::mojom::BatAdsLedgerFactory>() {
+#if !BUILDFLAG(IS_ANDROID)
+  return sandbox::mojom::Sandbox::kNoSandbox;
+#else
+  return sandbox::mojom::Sandbox::kUtility;
+#endif  // !BUILDFLAG(IS_ANDROID)
+}
+
+namespace combined_utility {
+
+namespace {
+
+scoped_refptr<ServiceInstance> MakeBatAdsLedgerFactoryInstance() {
+  mojo::Remote<combined_utility::mojom::BatAdsLedgerFactory>
+      bat_ads_ledger_service;
+  content::ServiceProcessHost::Launch(
+      bat_ads_ledger_service.BindNewPipeAndPassReceiver(),
+      content::ServiceProcessHost::Options()
+          .WithDisplayName("TODO_NAME")
+          .Pass());
+  return base::MakeRefCounted<ServiceInstance>(
+      std::move(bat_ads_ledger_service));
+}
+
+}  // namespace
+
+ServiceInstance::ServiceInstance(
+    mojo::Remote<combined_utility::mojom::BatAdsLedgerFactory> interface_remote)
+    : interface_remote_(std::move(interface_remote)) {}
+
+ServiceInstance::~ServiceInstance() = default;
+
+scoped_refptr<ServiceInstance>
+CombinedUtilityService::MakeStrongBatAdsLedgerFactory() {
+  auto strong_factory_ref =
+      base::WrapRefCounted(bat_ads_ledger_factory_weak_.get());
+  if (!strong_factory_ref ||
+      !strong_factory_ref->interface_remote().is_bound()) {
+    strong_factory_ref = MakeBatAdsLedgerFactoryInstance();
+    bat_ads_ledger_factory_weak_ = strong_factory_ref->AsWeakPtr();
+  }
+  return strong_factory_ref;
+}
+
+InterfaceHolder<bat_ledger::mojom::BatLedgerService>
+CombinedUtilityService::MakeBatLedgerService() {
+  auto strong_factory_ref = MakeStrongBatAdsLedgerFactory();
+  mojo::Remote<bat_ledger::mojom::BatLedgerService> bat_ledger_service;
+  strong_factory_ref->interface_remote()->MakeBatLedgerService(
+      bat_ledger_service.BindNewPipeAndPassReceiver());
+  return InterfaceHolder(std::move(bat_ledger_service),
+                         std::move(strong_factory_ref));
+}
+
+InterfaceHolder<bat_ads::mojom::BatAdsService>
+CombinedUtilityService::MakeBatAdsService() {
```

Comment:
I don't think this is something we want to do in general like I mentioned in the other comment, but we definitely shouldn't have ads/ledger specific methods here if this is supposed to be a generic service. This could all be handled with templates

Reply:
although actually if we do keep this, I think it should just move into the rewards component. You can have more than one keyed service there and ads already has a dependency on rewards